### PR TITLE
Fix multi reset bugs in delay_queue

### PR DESCRIPTION
**Pull request by zaharidichev**
## Motivation
Fixes a bug that causes problems when doing multiple resets in delay queue.
## Solution
We currently need to store both the delay and the key of the first expiring item, in order to know whether we should reset the delay of the first expiring item or not

### CODE REVIEW

**Comment by zaharidichev** [comment]
@Ralith, probably deserves a look from you.

File: tokio-timer/src/delay_queue.rs [hunk 1]
```diff
@@ -507,24 +513,27 @@ impl<T> DelayQueue<T> {
...
+            // 2. The item to expire first is the same as the one
+            // we are resetting AND the instant to which we are resetting
+            // is further out in time...
+
+            let new_exp = self.start + Duration::from_millis(when);
+            let reset_item_expires_first = new_exp < delay.deadline();
+            let reset_item_is_current_first_exp = index == key.index;
+            let reset_to_later_exp = new_exp > delay.deadline();
+            if reset_item_expires_first || (reset_item_is_current_first_exp && reset_to_later_exp) {
+                delay.reset(new_exp);
```
Comment (Ralith) [inline tokio-timer/src/delay_queue.rs]: If there are two elements in the queue, and the initially-sooner one is reset to be later than the other, won't this result in the delay being set incorrectly? Perhaps `new_exp` should be taken from `wheel.poll_at()`?

Comment (zaharidichev) [reply tokio-timer/src/delay_queue.rs]: Yes.. you are right. And in this case we need to be able to be able to exchange the key as well to be the key of the new "about to expire" item..  Hmm this tracking of first to expire + resetting the timeouts introduces some headache...

Comment (zaharidichev) [reply tokio-timer/src/delay_queue.rs]: Will have to give it a bit more thought to make sure I cover all cases

**Comment by zaharidichev** [comment]
@Ralith I think this should do it.

File: tokio-timer/src/delay_queue.rs [hunk 2]
```diff
@@ -850,4 +842,4 @@ impl<T> Expired<T> {
     pub fn into_inner(self) -> T {
         self.data
     }
-}
+}
```
Comment (Ralith) [inline tokio-timer/src/delay_queue.rs]: Unintended whitespace change?

**Pull request merged**